[fix] Add AbortController for async upload operations

sfc-gh-nbellante opened the pull request:
## Describe your changes
This pull request improves the robustness of the `AudioInput` component by adding support for aborting ongoing audio uploads. This ensures that if a new upload is started or the component is unmounted, any previous or ongoing upload is properly cancelled, preventing race conditions and unnecessary state updates. (And the "tried to update state on an unmounted component" error.)

<!-- If it's a visual change, please include a screenshot or video! -->

## GitHub Issue Link (if applicable)

No filed issue.

## Testing Plan

We could I guess try to repro this behavior in the tests, but since this is a preventative measure, I don't think the ROI is worth adding a test for this.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
fix: Add AbortController for async upload operations

- Add uploadAbortControllerRef to track active uploads
- Cancel previous uploads when starting new ones
- Check abort signal before state updates
- Clean up abort controller on component unmount

This prevents state updates after unmount and properly cancels
in-flight upload operations when the component unmounts or a new
recording starts.

sfc-gh-nbellante commented:
* **#12539** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12539?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12539?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12542** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12542?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12541** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12541?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12540** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12540?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -230,15 +257,21 @@ const AudioInput: React.FC<Props> = ({
           }
         } catch {
-          setIsError(true)
+          if (!abortController.signal.aborted) {
+            setIsError(true)
+          }
```
Meh, idk if I agree with this.

sfc-gh-nbellante commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -230,15 +257,21 @@ const AudioInput: React.FC<Props> = ({
             setDeleteFileUrl(upload.fileUrl.deleteUrl)
           }
         } catch {
-          setIsError(true)
+          if (!abortController.signal.aborted) {
+            setIsError(true)
+          }
         } finally {
           if (notNullOrUndefined(widgetFormId))
             widgetMgr.setFormsWithUploadsInProgress(new Set())
-          setIsUploading(false)
+          if (!abortController.signal.aborted) {
+            setIsUploading(false)
+          }
```
The widget manager lives outside of the scope of this component, so I think this should be fine to have this check unguarded.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/AudioInput/AudioInput.tsx:
```diff
@@ -142,9 +142,19 @@ const AudioInput: React.FC<Props> = ({
   const recordPluginHandlersRef = useRef<{
     handleRecordProgress?: (time: number) => void
   }>({})
+  const uploadAbortControllerRef = useRef<AbortController | null>(null)
```
**suggestion:** Please add tests for this behavior.

sfc-gh-nbellante committed:
Fix AbortController implementation for AudioInput

- Pass AbortSignal from AudioInput to uploadFiles utility
- Update uploadFiles to accept and forward AbortSignal to uploadClient
- Add E2E tests to verify rapid re-recordings work without race conditions
- Add test for blob URL cleanup to prevent memory leaks

This ensures that when a new recording starts, the previous upload is properly
aborted, preventing race conditions where an old recording could overwrite a
newer one.

sfc-gh-nbellante committed:
Improve E2E tests to avoid unnecessary wait_for_timeout

- Remove wait_for_timeout(2000) and replace with proper expect conditions
- Use expect().to_be_visible() for waiting on elements (built-in waiting)
- Keep wait_for_timeout only for actual recording duration (unavoidable)
- Add proper wait conditions between recordings using expect()
- Follow Streamlit E2E test guide recommendation to avoid wait_for_timeout

sfc-gh-nbellante committed:
test: Improve blob URL test reliability with wait_until pattern

- Use wait_until pattern for async checks before assertions
- Keep assert statements for non-DOM values as per codebase conventions
- Add comments explaining why asserts are acceptable for JavaScript evaluations
- This aligns with patterns used in st_heading_test.py and other tests
